src/app/api/webhooks/stripe/route.ts

```diff
@@ -0,0 +1,119 @@
+import Stripe from "stripe"
+import { stripe } from "@/services/stripe"
+import { env } from "@/env"
+import pino from "pino"
+import { getLogger } from "@/logging/log-util"
+import { db } from "@/services/db"
+import { MembershipStatus } from "@prisma/client"
+
+const logger = getLogger("stripe-webhook")
+
+export async function POST(req: Request) {
+  const body = await req.text()
+  const signature = req.headers.get("Stripe-Signature") as string
+
+  let event: Stripe.Event
+
+  try {
+    event = stripe.webhooks.constructEvent(
+      body,
+      signature,
+      env.STRIPE_WEBHOOK_SECRET,
+    )
+  } catch (error: any) {
+    return new Response(`Webhook Error: ${error.message}`, { status: 400 })
+  }
+
+  const session = event.data.object as Stripe.Subscription
+
+  if (event.type === "customer.subscription.updated") {
+    return await updateSubscription(session)
+  }
+
+  if (event.type === "customer.subscription.deleted") {
+    return await cancelSubscription(session)
+  }
+
+  return new Response(`Nothing to do: ${event.type}`, { status: 200 })
+}
+
+async function cancelSubscription(
+  subscription: Stripe.Subscription,
+): Promise<Response> {
+  logger.info(`Subscription canceled: ${subscription.id}`)
+  let membership = await db.membership.findUnique({
+    where: {
+      stripeSubscriptionId: subscription.id,
+    },
+  })
+  if (!membership) {
```

Comment:
no auth or double check for cancel subscription?

Reply:
This is a stripe webhook so I am just reporting the fact that the user has decided to cancel its subscription

Reply:
How we can be sure that the request is starting from stripe?

Reply:
The check is done by using a webhook secret created by Stripe itself.

https://github.com/Schrodinger-Hat/peoplez/blob/d71af17c061368088b6ed3f8e1c990c7f3d1dd3a/src/app/api/webhooks/stripe/route.ts#L12-L25